overlord/snapstate/autorefresh.go

```diff
@@ -994,13 +995,43 @@ func incrementSnapRefreshFailures(st *state.State, snapsup *SnapSetup) error {
 			LastFailureTime: timeNow(),
 		}
 	}
+	snapst.RefreshFailures.LastFailureSeverity = severity
 	Set(st, snapsup.InstanceName(), &snapst)
 
 	delay := computeSnapRefreshRemainingDelay(snapst.RefreshFailures).Round(time.Hour)
 	logger.Noticef("snap %q auto-refresh to revision %s has failed, next auto-refresh attempt will be delayed by %v hours", snapsup.InstanceName(), snapsup.Revision(), delay.Hours())
 	return nil
 }
 
+func computeSnapRefreshFailureSeverity(chg *state.Change, unlinkTask *state.Task, snapName string) snap.RefreshFailureSeverity {
+	laneTasks := chg.LaneTasks(unlinkTask.Lanes()...)
+	// Look for a tasks marked as a restart boundary.
+	for _, t := range laneTasks {
+		// If a task is found in an Undone state with its restart boundary in the "do"
+		// direction then this indicates that the auto-refresh failed after a reboot
+		// and we should be more aggressive with the backoff delay applied to it to
+		// decrease devices' downtime due to bad snap revisions.
+		if t.Status() != state.UndoneStatus || !restart.TaskIsRestartBoundary(t, restart.RestartBoundaryDirectionDo) {
+			continue
+		}
+		// let's double check task is for the guilty snap since the refresh
+		// change might be running in a transaction and have all snaps'
+		// refresh tasks in the same lane.
+		snapsup, err := TaskSnapSetup(t)
+		if err != nil {
+			logger.Debugf("internal error: failed to get snap associated with task %s: %v", t.ID(), err)
+			continue
+		}
+		if snapsup.InstanceName() != snapName {
```

Comment:
shouldn't we check this first?

Reply:
I am not sure it would make a logical difference, no? My intent was to do the cheaper checks first (checking the status and restart-boundry) to make it faster.

Reply:
sorry, I reviewed quickly and didn't connect this to the TaskSnapSetup which is indeed "expensive"